core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -0,0 +1,243 @@
+		uriParts, err := url.Parse(reference)
+		if err != nil {
+			return false
+		} else if uriParts.Scheme == "gs" {
+			return true
+		}
+	}
+	return false
+}
+
+// Upload uploads a file to the server
+func (ft *GCSFileTransfer) Upload(task *Task) error {
+	ft.logger.Debug("gcs file transfer: uploading file", "path", task.Path, "url", task.Url)
+
+	return nil
+}
+
+// Download downloads a file from the server
+func (ft *GCSFileTransfer) Download(task *Task) error {
+	ft.logger.Debug("gcs file transfer: downloading file", "path", task.Path, "url", task.Url, "ref", task.Reference)
+	if task.Reference == nil {
+		err := fmt.Errorf("gcs file transfer: download: reference is nil")
+		ft.logger.CaptureError(err)
+		return err
+	}
+	reference := *task.Reference
+
+	// Parse the reference path to get the scheme, bucket, and object
+	uriParts, err := url.Parse(reference)
+	if err != nil {
+		ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error parsing reference %s: %v", reference, err))
+		return err
+	}
+	if uriParts.Scheme != "gs" {
+		err := fmt.Errorf("gcs file transfer: download: invalid gsutil URI %s", reference)
+		ft.logger.CaptureError(err)
+		return err
+	}
+	bucketName := uriParts.Host
+	bucket := ft.client.Bucket(bucketName)
+	objectName := strings.TrimPrefix(uriParts.Path, "/")
+
+	switch {
+	case task.Digest == reference:
+		// If not using checksum, get all objects under the reference
+		var objectNames []string
+		query := &storage.Query{Prefix: objectName}
+		it := bucket.Objects(ft.ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error iterating through gcs bucket %s with prefix %s for reference %s: %v", bucketName, objectName, reference, err))
+				return err
+			}
+			objectNames = append(objectNames, objAttrs.Name)
+		}
+
+		g := new(errgroup.Group)
+		g.SetLimit(maxWorkers)
+		for _, name := range objectNames {
+			objName := name // for closure in the goroutine
+			g.Go(func() error {
+				object := bucket.Object(objName)
+				localPath := getDownloadFilePath(objName, objectName, task.Path)
+				err := ft.DownloadFile(object, localPath)
+				if err != nil {
+					// this file is probably a folder, skip to avoid an unnecessary error
+					if isDir(reference, task.Size) {
+						ft.logger.Debug("gcs file transfer: download: skipping reference because it seems to be a folder", "reference", reference)
+						return nil
+					}
+					ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error when downloading file for reference %s: %v", reference, err))
+					return err
+				}
+				return nil
+			})
+		}
+
+		// Wait for all of the go routines to complete and return an error if any errored out
+		if err := g.Wait(); err != nil {
+			return err
+		}
+	case task.VersionId != nil:
+		object := bucket.Object(objectName).Generation(int64(task.VersionId.(float64)))
```

Comment:
Since we're doing type assertion here, if task.VersionId is not of type float64 it will cause a runtime panic. Consider doing this:

```go
func safeConvertToInt64(value interface{}) (int64, error) {
	floatVal, ok := value.(float64)
	if !ok {
		return 0, fmt.Errorf("value is not a float64: %v", value)
	}
	return int64(floatVal), nil
}
```

and can then use it like this:

```go
versionId, err := safeConvertToInt64(task.VersionId)
	if err != nil {
		log.Fatalf("failed to convert VersionId: %v", err)
	}
```

Reply:
good point, thanks for catching!